Split torch_ucc to torch_ucc and torch_ucc_oss

zasdfgbnm opened the pull request:
These two libraries have different init files: `torch_ucc` will build with `torch_ucc_init.cpp` and `torch_ucc_oss` will build with `torch_ucc_init_oss.cpp`

zasdfgbnm commented:
Ready for review

pallab-zz commented:
Since this is about integration with OSS/conda, we can move the suggested changes to a new file such as torch_ucc_init_oss.* and move the existing removals to torch_ucc_init.* - see suggestions

pallab-zz commented on include/torch_ucc.hpp:
```diff
@@ -11,7 +11,6 @@
 #pragma once
 
 #include "torch_ucc_comm.hpp"
-#include <torch/python.h>
```
move to torch_ucc_init.cpp

pallab-zz replied:
move to src/torch_ucc_init.hpp

pallab-zz commented on include/torch_ucc.hpp:
```diff
@@ -20,8 +19,6 @@
 #include <thread>
 #include <vector>
 
-#include <pybind11/chrono.h>
```
move to src/torch_ucc_init.hpp

pallab-zz commented on include/torch_ucc.hpp:
```diff
@@ -297,13 +294,6 @@ template <typename PreProcess, typename PostProcess>
       int size,
       const std::chrono::duration<float>& timeout);
 
-  static void ProcessGroupUCCConstructor() __attribute__((constructor)) {
```
move this function to src/torch_ucc_init.cpp

pallab-zz commented on setup.py:
```diff
@@ -32,7 +32,8 @@
 plugin_sources      = ["src/torch_ucc.cpp",
-                       "src/torch_ucc_comm.cpp"]
+                       "src/torch_ucc_comm.cpp",
+                       "src/torch_ucc_python.cpp"]
```
rename to src/torch_ucc_init_oss.cpp

pallab-zz commented on src/torch_ucc.cpp:
```diff
@@ -1253,8 +1253,4 @@ void ProcessGroupUCC::initComm(c10::Device dev) {
 #endif
 }
 
-PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
```
move to src/torch_ucc_init.cpp

pallab-zz commented on src/torch_ucc_python.cpp:
```diff
@@ -0,0 +1,12 @@
+#include <torch/python.h>
```
move this file to src/torch_ucc_init_oss.cpp

zasdfgbnm commented:
Hi @pallab-zz, thanks for the review! I like these filenames you suggest, so I renamed `torch_ucc_python.cpp` to `torch_ucc_init.cpp` and added a new empty `torch_ucc_init_oss.cpp` which will be the home for the `createProcessGroupUCCForNCCL` for my next PR https://github.com/facebookresearch/torch_ucc/pull/57.

I am not sure if I understand you correctly. Are you saying that you want to keep the `ProcessGroupUCC::ProcessGroupUCCConstructor` and move the `PYBIND11_MODULE` to `torch_ucc_init_oss.cpp`? I don't see why we should do so, because this PR doesn't have any user-visible change for `torch_ucc` users, and `PYBIND11_MODULE` is only for these users, PyTorch NCCL integration doesn't need the `PYBIND11_MODULE`. Also, `ProcessGroupUCC::ProcessGroupUCCConstructor` is a class member, and moving it to a different file doesn't change the behavior. We have to remove `ProcessGroupUCC::ProcessGroupUCCConstructor` and move its work to `PYBIND11_MODULE`.

pallab-zz commented:
Hi @zasdfgbnm 
to work as a attribute constructor - there is no requirement to be a class member - the only thing it refers to from the ProcessGroupUCC class is the static member method here 
>>    register_backend("ucc", py::cpp_function(createProcessGroupUCC));
so when you move the function out - make it a standard static function - and replace the above line with
>>     register_backend("ucc", py::cpp_function(ProcessGroupUCC::createProcessGroupUCC));
In the oss variant you do not link with this file : torch_ucc_init.cpp

zasdfgbnm commented:
@pallab-zz I have modified my code according to your review.

zasdfgbnm commented on src/torch_ucc_init_oss.cpp:
```diff
@@ -0,0 +1,4 @@
+#include <torch/python.h>
+
+PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
```
I still need this so that I can get the filename of the shared object with
```
python -c "import inspect; import torch; import torch_ucc_oss; print(inspect.getfile(torch_ucc_oss))"
```

pallab-zz replied:
Thanks much @zasdfgbnm - pl give me some time to test it - will let u know soon

pallab-zz commented on setup.py:
```diff
@@ -40,11 +40,19 @@
+if '--oss' in sys.argv:
+  sys.argv.remove('--oss')
+  plugin_sources += ["src/torch_ucc_init_oss.cpp"]
+  name = "torch_ucc_oss"
```
We do not need to change the name to torch_ucc_oss - the name can and should remain same. regardless of which cpp file is used.

zasdfgbnm replied:
fixed

pallab-zz commented on setup.py:
```diff
@@ -40,11 +40,19 @@
+  name = "torch_ucc_oss"
+else:
+  plugin_sources += ["src/torch_ucc_init.cpp"]
+  name = "torch_ucc"
```
this name is what should be used regardless of whether oss or not

zasdfgbnm replied:
fixed

pallab-zz commented:
Hi @zasdfgbnm - thanks much - i tested a small one - and it works for the non-oss case - i am going to assume you will test the oss case and hence i will approve

pallab-zz commented:
actually may i req to wait a bit - since i see @kingchc has pulled it - would like to give him a chance to chime in if that is okey ..

kingchc reviewed (approved):
LGTM. I l also tested non-oss part and it works fine. CI tests should cover the oss part, and it looks passing.
